File: app/src/test/java/no/nav/tilbakemeldingsmottak/validators/KlassifiserServiceklageValidatorTest.java
package no.nav.tilbakemeldingsmottak.validators;

import no.nav.tilbakemeldingsmottak.exceptions.ClientErrorException;
import no.nav.tilbakemeldingsmottak.model.HentSkjemaResponse;
import no.nav.tilbakemeldingsmottak.model.KlassifiserServiceklageRequest;
import no.nav.tilbakemeldingsmottak.rest.serviceklage.validation.KlassifiserServiceklageValidator;
import org.junit.jupiter.api.Test;

import static no.nav.tilbakemeldingsmottak.TestUtils.*;
import static no.nav.tilbakemeldingsmottak.serviceklage.ServiceklageConstants.KANAL_MUNTLIG_ANSWER;
import static org.junit.Assert.assertTrue;
import static org.junit.jupiter.api.Assertions.assertThrows;

class KlassifiserServiceklageValidatorTest {

    private final KlassifiserServiceklageValidator klassifiserServiceklageValidator = new KlassifiserServiceklageValidator();
    private KlassifiserServiceklageRequest klassifiserServiceklageRequest;

    @Test
    void happyPathServiceklage() {
        klassifiserServiceklageRequest = createKlassifiserServiceklageRequest();
        klassifiserServiceklageValidator.validateRequest(klassifiserServiceklageRequest, createHentSkjemaResponse());
    }

    @Test
    void happyPathAnnet() {
        klassifiserServiceklageRequest = createKlassifiserServiceklageRequestIkkeServiceklage();
        klassifiserServiceklageValidator.validateRequest(klassifiserServiceklageRequest, createHentSkjemaResponse());
    }

    @Test
    void happyPathForvaltningsklage() {
        klassifiserServiceklageRequest = createKlassifiserServiceklageRequestForvaltningsklage();
        klassifiserServiceklageValidator.validateRequest(klassifiserServiceklageRequest, createHentSkjemaResponse());
    }

    @Test
    void shouldThrowExceptionIfChoiceNotValid() {
        klassifiserServiceklageRequest = createKlassifiserServiceklageRequest();
        klassifiserServiceklageRequest.setTEMA("Ugyldig valg");

        Exception thrown = assertThrows(ClientErrorException.class,
                () -> klassifiserServiceklageValidator.validateRequest(klassifiserServiceklageRequest, createHentSkjemaResponse()));
        assertTrue(thrown.getMessage().contains("Innsendt svar på spørsmål med id=TEMA er ikke gyldig"));
    }

    @Test
    void shouldThrowExceptionIfTextIsBlank() {
        klassifiserServiceklageRequest = createKlassifiserServiceklageRequest();
        klassifiserServiceklageRequest.setAARSAK("");

        Exception thrown = assertThrows(ClientErrorException.class,
                () -> klassifiserServiceklageValidator.validateRequest(klassifiserServiceklageRequest, createHentSkjemaResponse()));
        assertTrue(thrown.getMessage().contains("Innsendt svar på spørsmål med id=AARSAK er ikke gyldig"));
    }

    @Test
    void shouldThrowExceptionIfDateIsInvalid() {
        klassifiserServiceklageRequest = createKlassifiserServiceklageRequest();
        klassifiserServiceklageRequest.setFREMMETDATO("123");

        Exception thrown = assertThrows(ClientErrorException.class,
                () -> klassifiserServiceklageValidator.validateRequest(klassifiserServiceklageRequest, createHentSkjemaResponse()));
        assertTrue(thrown.getMessage().contains("Innsendt svar på spørsmål med id=FREMMET_DATO er ikke gyldig"));
    }

    @Test
    void shouldThrowExceptionIfDefaultAnswerDoesntMatch() {
        klassifiserServiceklageRequest = createKlassifiserServiceklageRequest();
        HentSkjemaResponse hentSkjemaResponse = createHentSkjemaResponseWithDefaultAnswers();
        klassifiserServiceklageRequest.setKANAL(KANAL_MUNTLIG_ANSWER);

        Exception thrown = assertThrows(ClientErrorException.class,
                () -> klassifiserServiceklageValidator.validateRequest(klassifiserServiceklageRequest, hentSkjemaResponse));
        assertTrue(thrown.getMessage().contains("Innsendt svar på spørsmål med id=KANAL matcher ikke svar i database"));
    }

}

File: app/src/test/kotlin/no/nav/tilbakemeldingsmottak/validators/KlassifiserServiceklageValidatorTest.kt
package no.nav.tilbakemeldingsmottak.validators

import no.nav.tilbakemeldingsmottak.TestUtils.TestUtils.createHentSkjemaResponse
import no.nav.tilbakemeldingsmottak.TestUtils.TestUtils.createHentSkjemaResponseWithDefaultAnswers
import no.nav.tilbakemeldingsmottak.domain.ServiceklageConstants.KANAL_MUNTLIG_ANSWER
import no.nav.tilbakemeldingsmottak.exceptions.ClientErrorException
import no.nav.tilbakemeldingsmottak.model.HentSkjemaResponse
import no.nav.tilbakemeldingsmottak.model.KlassifiserServiceklageRequest
import no.nav.tilbakemeldingsmottak.rest.serviceklage.validation.KlassifiserServiceklageValidator
import no.nav.tilbakemeldingsmottak.util.builders.KlassifiserServiceklageRequestBuilder
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

internal class KlassifiserServiceklageValidatorTest {
    private val klassifiserServiceklageValidator = KlassifiserServiceklageValidator()
    private var klassifiserServiceklageRequest: KlassifiserServiceklageRequest =
        KlassifiserServiceklageRequestBuilder().build()

    @Test
    fun happyPathServiceklage() {
        klassifiserServiceklageRequest = KlassifiserServiceklageRequestBuilder().build()
        klassifiserServiceklageValidator.validateRequest(klassifiserServiceklageRequest, createHentSkjemaResponse())
    }

    @Test
    fun happyPathAnnet() {
        klassifiserServiceklageRequest = KlassifiserServiceklageRequestBuilder().asNotServiceklage().build()
        klassifiserServiceklageValidator.validateRequest(klassifiserServiceklageRequest, createHentSkjemaResponse())
    }

    @Test
    fun happyPathForvaltningsklage() {
        klassifiserServiceklageRequest = KlassifiserServiceklageRequestBuilder().asForvaltningsklage().build()
        klassifiserServiceklageValidator.validateRequest(klassifiserServiceklageRequest, createHentSkjemaResponse())
    }

    @Test
    fun shouldThrowExceptionIfChoiceNotValid() {
        // Given
        klassifiserServiceklageRequest = KlassifiserServiceklageRequestBuilder().build(TEMA = "Ugyldig valg")

        // When
        val thrown: Exception = Assertions.assertThrows(ClientErrorException::class.java) {
            klassifiserServiceklageValidator.validateRequest(
                klassifiserServiceklageRequest,
                createHentSkjemaResponse()
            )
        }

        // Then
        assertTrue(thrown.message!!.contains("Innsendt svar på spørsmål med id=TEMA er ikke gyldig"))
    }

    @Test
    fun shouldThrowExceptionIfTextIsBlank() {
        // Given
        klassifiserServiceklageRequest = KlassifiserServiceklageRequestBuilder().build(AARSAK = "")

        // When
        val thrown: Exception = Assertions.assertThrows(ClientErrorException::class.java) {
            klassifiserServiceklageValidator.validateRequest(
                klassifiserServiceklageRequest,
                createHentSkjemaResponse()
            )
        }

        // Then
        assertTrue(thrown.message!!.contains("Innsendt svar på spørsmål med id=AARSAK er ikke gyldig"))
    }

    @Test
    fun shouldThrowExceptionIfDateIsInvalid() {
        // Given
        klassifiserServiceklageRequest = KlassifiserServiceklageRequestBuilder().build(FREMMET_DATO = "123")

        // When
        val thrown: Exception = Assertions.assertThrows(ClientErrorException::class.java) {
            klassifiserServiceklageValidator.validateRequest(
                klassifiserServiceklageRequest,
                createHentSkjemaResponse()
            )
        }

        // Then
        assertTrue(thrown.message!!.contains("Innsendt svar på spørsmål med id=FREMMET_DATO er ikke gyldig"))
    }

    @Test
    fun shouldThrowExceptionIfDefaultAnswerDoesntMatch() {
        // Given
        klassifiserServiceklageRequest = KlassifiserServiceklageRequestBuilder().build(KANAL = KANAL_MUNTLIG_ANSWER)
        val hentSkjemaResponse: HentSkjemaResponse = createHentSkjemaResponseWithDefaultAnswers()

        // When
        val thrown: Exception = Assertions.assertThrows(ClientErrorException::class.java) {
            klassifiserServiceklageValidator.validateRequest(
                klassifiserServiceklageRequest,
                hentSkjemaResponse
            )
        }

        // Then
        assertTrue(thrown.message!!.contains("Innsendt svar på spørsmål med id=KANAL matcher ikke svar i database"))
    }
}